lib/dry/types/compiler.rb

```diff
@@ -11,24 +13,30 @@ def call(ast)
         visit(ast)
       end
 
-      def visit(node, *args)
-        send(:"visit_#{node[0]}", node[1], *args)
+      def visit(node)
+        method, *args = extract_method_and_args(node)
+        send(:"visit_#{method}", *args)
       end
 
       def visit_constructor(node)
-        primitive, fn = node
+        definition, fn = node
+        primitive = visit(definition)
         Types::Constructor.new(primitive, &fn)
       end
 
-      def visit_type(node)
-        type, args = node
-        meth = :"visit_#{type.tr('.', '_')}"
+      def visit_safe(node)
+        method, *args = node
+        Types::Safe.new(send(:"visit_#{method}", *args))
+      end
 
-        if respond_to?(meth) && args
-          send(meth, args)
-        else
-          registry[type]
-        end
+      def visit_definition(*node)
+        primitive, meta = node
+        definition = if registry.registered?(primitive)
+                       registry[primitive]
+                     else
+                       Definition.new(primitive)
+                     end
+        meta.any? ? definition.meta(meta) : definition
```

Comment:
I think even if you don't have any `meta` by default it would by an empty hash, since we added meta as part of the `ast` https://github.com/dry-rb/dry-types/pull/178/files#diff-58177960b3ed930fac3ecb590a2d3f35R108